lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -338,6 +465,10 @@ class LiftArrayAllocPass
       RewritePatternSet patterns(ctx);
       patterns.insert<AllocaPattern>(ctx, domInfo, funcName, module);
       patterns.insert<ComplexCreatePattern>(ctx);
+      patterns.insert<FloatExtendPattern>(ctx);
+      patterns.insert<FloatTruncatePattern>(ctx);
+      patterns.insert<ComplexRePattern>(ctx);
+      patterns.insert<ComplexImPattern>(ctx);
```

Comment:
These patterns seem like they should already be part of canonicalization. Are they? If not, is it because we're using an old version of MLIR?

If MLIR isn't canonicalizing these, we should use cc::CastOp and add these patterns to its canonicalization patterns. This functionality isn't conceptually part of lifting array values.

Reply:
The Trunc and Ext are definitely not handled (my new tests will fail without them) but I Re and IM pattern was my try to const prop conversion to the correct simulation precision (which is not working yet, plan to address it in another PR).

Reply:
I'm getting a different result regarding this. I'll send you the test.

Reply:
> we should use cc::CastOp
This would be a change in ConvertExpr which I would like to do separately

Reply:
Will add those to canonicalizer, among other things, in a separate PR, as discussed.